Update Transaction Details breakdown: rename labels and add fee breakdown tooltip when disputed

Jinksi opened the pull request:
Fixes #6974

> **Warning**
> This PR is **NOT** behind a feature flag


#### Changes proposed in this Pull Request

This PR updates the Transaction Details screen with the following changes:
* Update transaction breakdown labels:
	* ~Refunded~ → Deducted
		* only for disputed transactions where a dispute fee has been applied and not reversed
	* ~Fee~ → Fees
	* Change style to uppercase
* A click tooltip that only shows if there is a Dispute fee in addition to the transaction fee. Tooltip content includes fee breakdown:
	* Transaction fee
	* Dispute fee
	* Total fee

**Note:** HorizontalList uppercase labels were updated in #7126 

**Before:** (`develop`)
<img width="1032" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/3147296/b565cf55-30be-4d12-a0ce-1e6c140137e2">


**After:**
<img width="1033" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/3147296/8d8e4e2f-bb64-48f4-adcd-7578b9acbb17">

Fees tooltip content, multi-currency enabled:
<img width="1056" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/3147296/dad874aa-ceeb-4fd8-b1b4-7b4320f37b3c">


**Design:**
<img width="1001" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/3147296/2644bd8d-78a2-4fc6-91e6-ccd79f5bf4f3">


Additionally, some other code improvements:
* Add the `reporting_category` property to the `BalanceTransaction` type.
* Fix a `ClickTooltip` bug preventing the `className` prop from being applied to the tooltip content element.

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Create a disputed transaction using the card `4000000000000259`.
* Navigate to this disputed via the Payments → Transactions list screen.
* Ensure the Fees tooltip renders, along with the correct fee breakdown.
	* Ensure the **Transaction fee** matches the **Fee** total found on the timeline further down this transaction details screen:
		<img width="400" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/3147296/6947f200-6b99-4ab8-a4cd-3893e40df300">
* View a non-disputed transaction details screen.
* Ensure the Fees tooltip does not render.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

Jinksi committed:
Change 'Refunded' label to 'Deducted'

Jinksi committed:
Change 'Fee' label to 'Fees'

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7118` or branch name `update/6974-transaction-details-breakdown-labels-and-fees` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=update/6974-transaction-details-breakdown-labels-and-fees) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 4c24b59499c376a83af0f098128b9d86a38a8973
- Build time: 2023-09-20 21:00:30 UTC

_Note: the build is updated when a new commit is pushed to this PR._

Jinksi committed:
Add missing `reporting_category` to BalanceTransaction type

in order to filter balance transactions by 'dispute' and get the dispute fee

Jinksi committed:
Add getDisputeFee util function

Jinksi committed:
Fix for ClickTooltip className prop not being applied

Jinksi committed:
Add tooltip content and styles

Jinksi committed:
Add test for tooltip fee breakdown

Jinksi committed:
Merge branch 'develop' into update/6974-transaction-details-breakdown-labels-and-fees

Jinksi committed:
Reorder imports to improve readability

Jinksi committed:
Update list item title styles to uppercase, match design

Jinksi committed:
Fix vertical alignment for list items with multiple content elements

e.g. text & ClickTooltip

Jinksi committed:
Add empty changelog entry

Jinksi committed:
Merge branch 'update/horizontal-list-label-style-uppercase' into update/6974-transaction-details-breakdown-labels-and-fees

Jinksi committed:
Update breakdown styles to uppercase, matching design

Jinksi committed:
Increase font sizes to improve text readability

Jinksi committed:
Add blank changelog entry

Jinksi committed:
Merge branch 'develop' into update/6974-transaction-details-breakdown-labels-and-fees

Jinksi committed:
Merge branch 'develop' into update/6974-transaction-details-breakdown-labels-and-fees

Jinksi committed:
Merge branch 'develop' into update/6974-transaction-details-breakdown-labels-and-fees

Jinksi committed:
Merge branch 'develop' into update/6974-transaction-details-breakdown-labels-and-fees

shendy-a8c committed:
Merge branch 'develop' into update/6974-transaction-details-breakdown-labels-and-fees

shendy-a8c commented on changelog/update-6974-transaction-details-breakdown-labels-and-fees:
```diff
@@ -0,0 +1,5 @@
+Significance: patch
```
Super duper `nit`. I think it should be a `minor` as it's not a bug fix.
>   [patch] Backwards-compatible bug fixes.
  [minor] Added (or deprecated) functionality in a backwards-compatible manner.
  [major] Broke backwards compatibility in some way.

Jinksi replied:
Thanks, resolved in 754b6c3

shendy-a8c commented on changelog/update-6974-transaction-details-breakdown-labels-and-fees:
```diff
@@ -0,0 +1,5 @@
+Significance: patch
+Type: add
+Comment: Behind feature flag: update Transaction Details summary breakdown styles and fee breakdown tooltip.
```
Since this change is *not* behind a feature flag, this changelog should be updated and we need to introduce a new changelog entry.

I thought I read from the issue https://github.com/Automattic/woocommerce-payments/issues/6974 that it should be behind the feature flag but this PR's body states that it's not behind the feature flag.

The code does not indicate that it checks for the feature flag either.

Now I'm not sure where I read that this change should be behind the feature flag.

Anyway, IMO it does not need to be behind the feature flag.

Jinksi replied:
Yep, I've confused myself with the feature flag situation here. Thanks for spotting this!

I'll take a look and update here with the reasoning for/against.

Jinksi replied:
I've opted for not-behind-feature-flag for this issue because:
1. It is a relatively low-risk change.
2. It will benefit merchants immediately without relying on the rest of the project's work within the feature-flag.

Updated in 754b6c3

shendy-a8c reviewed (approved):
Diff looks good and every thing is functional.

Tested with multi currencies as well. Looks a bit weird seeing numbers in different currencies (especially one that is zero-decimal currency) summed up like this but I think it's OK.

![Screenshot 2023-09-11 at 03 42 21](https://github.com/Automattic/woocommerce-payments/assets/73803630/5aadec01-6f30-43fc-950d-e4bb79348c00).

Also, for the 'Refunded' to 'Deducted' change, it does not sound natural to me if the transaction does not have dispute and just a regular refunded transaction but it's just my personal opinion.

shendy-a8c commented on client/payment-details/summary/index.tsx:
```diff
@@ -177,6 +186,11 @@ const PaymentDetailsSummary: React.FC< PaymentDetailsSummaryProps > = ( {
 	const isFraudOutcomeReview = isOnHoldByFraudTools( charge, paymentIntent );
 
+	const disputeFee = charge.dispute && getDisputeFee( charge.dispute );
+	const transactionFee = charge.balance_transaction
+		? charge.balance_transaction.fee
+		: charge.application_fee_amount;
```
I don't understand this part. Could you explain?

Jinksi replied:
Good catch!

I understood the need for this approach by observing the following ternary statement introduced in the commit 7b4b9ba.

https://github.com/Automattic/woocommerce-payments/blob/648f0a1e1116e2f621ac7a0a69bb43bbded6b242/client/utils/charge/index.ts#L137-L151

From my understanding, the `balance_transaction` fee should exist for all captured payments. If a payment has been authorised but not captured, we can show the `application_fee_amount`, which should be identical.

I've added a comment with this reasoning in 648f0a1.

shendy-a8c replied:
Thank you for explaining that and the comment is really helpful.

shendy-a8c reviewed (dismissed):
Changelog change needed.

Jinksi commented:
Unrelated failing JS test being addressed in #7164.

Jinksi committed:
Merge branch 'develop' into update/6974-transaction-details-breakdown-labels-and-fees

Jinksi committed:
Updated non-feature-flag changelog entry

Jinksi commented:
> Also, for the 'Refunded' to 'Deducted' change, it does not sound natural to me if the transaction does not have dispute and just a regular refunded transaction but it's just my personal opinion.

@shendy-a8c it is an interesting rename for refunded transactions – I've asked for clarification here: p1694412278664569-slack-C03LA109BKM.

Jinksi committed:
Merge branch 'develop' into update/6974-transaction-details-breakdown-labels-and-fees

Jinksi commented on client/payment-details/summary/index.tsx:
```diff
@@ -229,7 +243,7 @@ const PaymentDetailsSummary: React.FC< PaymentDetailsSummaryProps > = ( {
 								<p>
 									{ `${ __(
-										'Refunded',
+										'Deducted',
```
⚠️ This should be renamed `Deducted` for disputed transactions only – not for refunded transactions. p1694412278664569-slack-C03LA109BKM

Jinksi replied:
Addressed in ee89414

Jinksi replied:
Added explicit tests for this logic in f0b9f52.

Jinksi committed:
Merge branch 'develop' into update/6974-transaction-details-breakdown-labels-and-fees

Jinksi committed:
Show "Deducted" fee label only for disputed transactions

Jinksi committed:
Add descriptive comment about auth/capture fee fallback

Jinksi committed:
Add explicit tests for Refunded/Deducted label logic

Jinksi committed:
Merge branch 'develop' into update/6974-transaction-details-breakdown-labels-and-fees

Jinksi commented:
> Tested with multi currencies as well. Looks a bit weird seeing numbers in different currencies 

@shendy-a8c, does the display of fees and currencies match what is listed in the transaction timeline?

I.e. is the dispute fee rendered in the store currency and the transaction fee in the charge currency?

Jinksi committed:
Don't render fee breakdown tooltip for `won` dispute

Jinksi commented on client/payment-details/summary/index.tsx:
```diff
@@ -177,6 +186,16 @@ const PaymentDetailsSummary: React.FC< PaymentDetailsSummaryProps > = ( {
 
+	const disputeFee =
+		charge.dispute &&
+		charge.dispute.status !== 'won' &&
```
Added these criteria prevent the fee tooltip rendering for `won` disputes.

<img width="1074" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/3147296/3de07f0b-1f5b-486f-a1a8-ba0a8456d531">

haszari replied:
I guess this logic could be moved to an accessor, e.g. `getDisputeFeeCharged`. What I read from this is that `getDisputeFee` is confusing. Could rename it `getDisputeBalanceTransaction` or something more technical/accurate.

Jinksi commented:
@nikkivias I have this PR active on the [Helix demo site](https://woo-delectably-sparkly-ninja.wpcomstaging.com/wp-admin/admin.php?page=wc-admin&path=/payments/transactions) if you'd like to take a look at the changes and offer feedback. 🙏

shendy-a8c committed:
Merge branch 'develop' into update/6974-transaction-details-breakdown-labels-and-fees

shendy-a8c commented:
> @shendy-a8c, does the display of fees and currencies match what is listed in the transaction timeline?

> I.e. is the dispute fee rendered in the store currency and the transaction fee in the charge currency?

The timeline, it shows both in customer currency and when it's converted (with the conversion rate) to the store currency.

![Screenshot 2023-09-12 at 20 38 24](https://github.com/Automattic/woocommerce-payments/assets/73803630/6e0b3c3c-804e-42d0-9adf-3d61b585f77e)

Here's my test transaction: p1694525759208469-slack-C03LA109BKM

Jinksi commented:
Marking as `needs author reply` while I test this across some more edge cases:
* single-deposit-currency stores with multi-currency disabled
* single-deposit-currency stores with multi-currency enabled
* multi-deposit-currency stores

Jinksi committed:
Merge branch 'develop' into update/6974-transaction-details-breakdown-labels-and-fees

Jinksi committed:
Ensure Transaction fee in tooltip is formatted in the fee currency

relevant for transactions where the charge currency differs from the fee balance transaction currency

Jinksi committed:
Rename transactionFee amount → fee for consistency

Jinksi commented:
Thanks for catching the issue with multi-currency transactions @shendy-a8c 🙏


I've addressed this in 8c9ef0e and b12f465 to now display the fee in the correct currency (the transaction fee's currency).

**Tooltip Transaction Fee** (kr41.60 in this case):
<img width="1056" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/3147296/52fb4499-e5c5-4ebd-8e38-b6b92bd9c086">


Now matches the **Transaction timeline fee total** (kr-41.60):

<img width="472" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/3147296/9fd66af5-c10a-4180-9fe4-bea7864e18b3">


[**See example multi-currency transaction on Helix Demo Site**](https://woo-delectably-sparkly-ninja.wpcomstaging.com/wp-admin/admin.php?page=wc-admin&path=/payments/transactions/details&id=pi_3Nq3cHCBu6Jj8nBr0lyI2OOs)

Jinksi committed:
Merge branch 'develop' into update/6974-transaction-details-breakdown-labels-and-fees

haszari committed:
Merge branch 'develop' into update/6974-transaction-details-breakdown-labels-and-fees

haszari commented on client/disputes/utils.ts:
```diff
@@ -69,3 +70,15 @@ export const isInquiry = ( dispute: Dispute | CachedDispute ): boolean => {
+	const disputeFee = dispute.balance_transactions.find(
+		( transaction ) => transaction.reporting_category === 'dispute'
+	);
+	return disputeFee;
```
Could this `find` ever return multiple items? Do we expect a single item returned from this function?

Jinksi replied:
I've added a description to the `Dispute` TS interface explaining what states the `balance_transactions` could represent, based on [the Stripe docs](https://stripe.com/docs/api/disputes/object#dispute_object-balance_transactions).

https://github.com/Automattic/woocommerce-payments/blob/764566d5e7d9a55933b432a6290d2c96d96f31c9/client/types/disputes.d.ts#L105-L110

I'm adjusting the `getDisputeFee` function to help make more sense of this.

Jinksi replied:
I've changed the `getDisputeFee` logic to clarify some things here.

https://github.com/Automattic/woocommerce-payments/blob/facb162feb86b6e0a4095364a9253caf06d427a3/client/disputes/utils.ts#L75-L114

@haszari do the comments and function names help to clarify the logic here?

haszari replied:
Crystal clear, nice work!

haszari commented on client/payment-details/summary/index.tsx:
```diff
@@ -246,13 +278,69 @@ const PaymentDetailsSummary: React.FC< PaymentDetailsSummaryProps > = ( {
+												</>
+											}
+										/>
+									) }
```
Suggest adding currency to all currency values for clarity. (We should probably review all recent changes to ensure we're consistently showing currency.)

This tooltip has similar information to the timeline, should we make them more consistent?

<img width="1210" alt="Screenshot 2023-09-18 at 10 27 09 AM" src="https://github.com/Automattic/woocommerce-payments/assets/4167300/40efeebb-880b-407b-aae3-c36e661f93f1">

For example:
- `Fee` => `Dispute fee` in timeline
- `$n USD deducted` (add `deducted` to line item in timeline)
- What's the $33.08 headline number in timeline 🤔?

Jinksi replied:
> What's the $33.08 headline number in timeline 🤔?

This I can answer: $18.08 USD disputed amount + $15 USD dispute fee.

Although, the current state of the UI doesn't make this clear. This could be moved to a new issue.

Related to this is the open PR that will change the Transaction Details headline amount to display in the store's currency instead of the customer's currency: #7159

haszari replied:
Sounds good to me – we can iterate on these details over time.

haszari commented on client/payment-details/summary/style.scss:
```diff
@@ -37,13 +37,35 @@
 			margin: 0.25rem 1rem 0 0;
+			text-transform: uppercase;
+			font-weight: 600;
+			font-size: 12px;
```
I find it hard to read with the data in the same all caps style as the headings/labels. @nikkivias what do you think about using different style to distinguish the labels from the data?

For example bold small caps for labels, regular weight for data. Mockup: 

<img width="270" alt="Screenshot 2023-09-18 at 10 45 36 AM" src="https://github.com/Automattic/woocommerce-payments/assets/4167300/f14297b4-5421-4b2f-8e85-44361c2316c9">

Jinksi replied:
From an engineering perspective, this is not a _completely_ straightforward change (these are all rendered as strings, e.g. `Deducted: $-18.08 USD`, so no separate elements to style).

Therefore, I'll leave this as is for this PR and we can iterate on this in a separate issue after the feedback session.

haszari replied:
Works for me. I also think we need to discuss this idea with design / other stakeholders before proceeding (though we could maybe prototype at appropriate priority).

haszari reviewed (approved):
This PR is working well and implements the design. 

I did some brief testing with a couple of disputes, tested with feature flag on and off. Also tested an undisputed transaction and no issues/regressions.

The styling of the transaction breakdown row doesn't feel right to me. The labels and the data use the same styles, which I find hard to read – the data should stand out, the labels are there to "onboard" and orient users to the values they need to find. This could be handled as follow up since this PR adds a useful breakdown tooltip. Keen for input from @nikkivias.

---

It got me thinking about a few things that we might want to look into (likely as follow up):

- Overall consistency / agreement on this page. Using the same terms/labels and breakdown in the header summary and the timeline. 
- Ensure we are using currency labels everywhere. There are some places these are left off and this is confusing. If a store only has 1 currency, no labels are needed, but if multicurrency is set up, every $$$ value should include currency IMO. 
- `NET` amount might need a breakdown tooltip too – but maybe there's a bigger confusion that we are trying to address with tooltips. I find it confusing that the deduction amount in the timeline includes the dispute fee but the summary NET / left number doesn't.
  - Actually that seems like a bug – `NET` should include all fees and deductions. 
- The first summary number is confusing without a label! I understand this is the purchase price after currency conversion.

Example transaction:

<img width="1210" alt="Screenshot 2023-09-18 at 10 27 09 AM" src="https://github.com/Automattic/woocommerce-payments/assets/4167300/ae748a5d-dcfb-4a65-978d-e7dd6e67ebb3">

Specific suggested changes (happy to log issues for this after discussion):

- Transaction summary `NET` should include dispute fee. In general, that line should add up to `NET`.
- All transaction timeline values should have currency (for multicurrency stores): loan repayment, disputed amount, dispute fee.

Consider:
- First value in summary line (purchase amount in store currency) - is this correct/confusing for disputes? 
- Timeline lists all items as `will be added from future deposit` `will be subtracted from future deposit`. Ideas:
  - Make timeline top-level sentence more descriptive, e.g.
    - `$2.73 USD deducted for loan repayment` (was `$2.73 will be subtracted from a future deposit.`)
    - `$33.23 USD deducted due to dispute` (was `$33.23 USD was deducted from your Jun 13, 2023 deposit.`)
- Some timeline items link to a real deposit, others like loan repayments seem to just say `Future deposit`. I wonder if we need to complicate timeline items with deposit info.

Jinksi committed:
Merge branch 'develop' into update/6974-transaction-details-breakdown-labels-and-fees

Jinksi committed:
Merge branch 'develop' into update/6974-transaction-details-breakdown-labels-and-fees

Jinksi committed:
Merge branch 'develop' into update/6974-transaction-details-breakdown-labels-and-fees

Jinksi committed:
Remove redundant snapshot file after merge

Jinksi committed:
Fix duplicate import

Jinksi committed:
Add Dispute interface balance_transactions description

Jinksi committed:
Improve getDisputeFee logic to ignore if deduction reversed

Jinksi committed:
Prefer usage of getDisputeFeeFormatted

and rename getDisputeDeductedBalanceTransaction to clarify intent

Jinksi committed:
Update getDisputeFeeFormatted description

Jinksi committed:
Add further code comments to explain logic

Jinksi committed:
Merge branch 'develop' into update/6974-transaction-details-breakdown-labels-and-fees

Jinksi committed:
Clarify comment about dispute balance_transaction count

Jinksi committed:
Don't render currency code in tooltip dispute fee
